File: packages/health_connector_hc_android/android/src/test/kotlin/com/phamtunglam/health_connector_hc_android/unit_tests/mappers/health_record_mappers/SkinTemperatureDeltaSeriesRecordMapperTest.kt
package com.phamtunglam.health_connector_hc_android.unit_tests.mappers.health_record_mappers

import androidx.health.connect.client.records.SkinTemperatureRecord
import androidx.health.connect.client.records.metadata.Metadata
import androidx.health.connect.client.units.Temperature
import androidx.health.connect.client.units.TemperatureDelta
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.DeviceTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.MetadataDto
import com.phamtunglam.health_connector_hc_android.pigeon.RecordingMethodDto
import com.phamtunglam.health_connector_hc_android.pigeon.SkinTemperatureDeltaSampleDto
import com.phamtunglam.health_connector_hc_android.pigeon.SkinTemperatureDeltaSeriesRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.SkinTemperatureMeasurementLocationDto
import io.kotest.matchers.shouldBe
import java.time.Instant
import java.time.ZoneOffset
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

@DisplayName("SkinTemperatureDeltaSeriesRecordMapper")
class SkinTemperatureDeltaSeriesRecordMapperTest {

    private companion object {
        const val TEST_ID = "skin-temp-delta-id"
        const val TEST_START_TIME = 1609459200000L
        const val TEST_END_TIME = 1609462800000L
        const val TEST_DELTA_TIME_1 = 1609459200000L
        const val TEST_DELTA_TIME_2 = 1609460100000L
        const val TEST_TEMPERATURE_DELTA_CELSIUS_1 = 0.5
        const val TEST_TEMPERATURE_DELTA_CELSIUS_2 = -0.3
        const val TEST_BASELINE_CELSIUS = 36.5
        val TEST_ZONE_OFFSET: ZoneOffset = ZoneOffset.ofHours(1)
    }

    @Test
    @DisplayName("GIVEN SkinTemperatureRecord → WHEN toDto → THEN converts correctly")
    fun whenRecordToDto_thenConvertsCorrectly() {
        // Given
        val record = SkinTemperatureRecord(
            startTime = Instant.ofEpochMilli(TEST_START_TIME),
            endTime = Instant.ofEpochMilli(TEST_END_TIME),
            startZoneOffset = TEST_ZONE_OFFSET,
            endZoneOffset = TEST_ZONE_OFFSET,
            metadata = Metadata.manualEntry(),
            deltas = listOf(
                SkinTemperatureRecord.Delta(
                    time = Instant.ofEpochMilli(TEST_DELTA_TIME_1),
                    delta = TemperatureDelta.celsius(TEST_TEMPERATURE_DELTA_CELSIUS_1),
                ),
                SkinTemperatureRecord.Delta(
                    time = Instant.ofEpochMilli(TEST_DELTA_TIME_2),
                    delta = TemperatureDelta.celsius(TEST_TEMPERATURE_DELTA_CELSIUS_2),
                ),
            ),
            baseline = Temperature.celsius(TEST_BASELINE_CELSIUS),
            measurementLocation = SkinTemperatureRecord.MEASUREMENT_LOCATION_WRIST,
        )

        // When
        val result = record.toDto()

        // Then
        result.startTime shouldBe TEST_START_TIME
        result.endTime shouldBe TEST_END_TIME
        result.startZoneOffsetSeconds shouldBe TEST_ZONE_OFFSET.totalSeconds.toLong()
        result.endZoneOffsetSeconds shouldBe TEST_ZONE_OFFSET.totalSeconds.toLong()
        result.samples.size shouldBe 2
        result.samples[0].time shouldBe TEST_DELTA_TIME_1
        result.samples[0].temperatureDeltaCelsius shouldBe TEST_TEMPERATURE_DELTA_CELSIUS_1
        result.samples[1].time shouldBe TEST_DELTA_TIME_2
        result.samples[1].temperatureDeltaCelsius shouldBe TEST_TEMPERATURE_DELTA_CELSIUS_2
        result.baselineCelsius shouldBe TEST_BASELINE_CELSIUS
        result.measurementLocation shouldBe SkinTemperatureMeasurementLocationDto.WRIST
    }

    @Test
    @DisplayName(
        "GIVEN SkinTemperatureDeltaSeriesRecordDto → WHEN toHealthConnect → THEN creates correctly",
    )
    fun whenDtoToHealthConnect_thenCreatesCorrectly() {
        // Given
        val dto = SkinTemperatureDeltaSeriesRecordDto(
            id = TEST_ID,
            startTime = TEST_START_TIME,
            endTime = TEST_END_TIME,
            startZoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            endZoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            metadata = MetadataDto(
                dataOrigin = "com.example.app",
                recordingMethod = RecordingMethodDto.MANUAL_ENTRY,
                deviceType = DeviceTypeDto.PHONE,
            ),
            samples = listOf(
                SkinTemperatureDeltaSampleDto(
                    time = TEST_DELTA_TIME_1,
                    temperatureDeltaCelsius = TEST_TEMPERATURE_DELTA_CELSIUS_1,
                ),
                SkinTemperatureDeltaSampleDto(
                    time = TEST_DELTA_TIME_2,
                    temperatureDeltaCelsius = TEST_TEMPERATURE_DELTA_CELSIUS_2,
                ),
            ),
            baselineCelsius = TEST_BASELINE_CELSIUS,
            measurementLocation = SkinTemperatureMeasurementLocationDto.FINGER,
        )

        // When
        val result = dto.toHealthConnect()

        // Then
        result.startTime shouldBe Instant.ofEpochMilli(TEST_START_TIME)
        result.endTime shouldBe Instant.ofEpochMilli(TEST_END_TIME)
        result.startZoneOffset shouldBe TEST_ZONE_OFFSET
        result.endZoneOffset shouldBe TEST_ZONE_OFFSET
        result.deltas.size shouldBe 2
        result.deltas[0].time shouldBe Instant.ofEpochMilli(TEST_DELTA_TIME_1)
        result.deltas[0].delta.inCelsius shouldBe TEST_TEMPERATURE_DELTA_CELSIUS_1
        result.deltas[1].time shouldBe Instant.ofEpochMilli(TEST_DELTA_TIME_2)
        result.deltas[1].delta.inCelsius shouldBe TEST_TEMPERATURE_DELTA_CELSIUS_2
        result.baseline?.inCelsius shouldBe TEST_BASELINE_CELSIUS
        result.measurementLocation shouldBe SkinTemperatureRecord.MEASUREMENT_LOCATION_FINGER
        result.metadata.id shouldBe TEST_ID
    }

    @Test
    @DisplayName(
        "GIVEN SkinTemperatureRecord with null baseline → WHEN toDto → THEN converts correctly",
    )
    fun whenRecordToDtoWithNullBaseline_thenConvertsCorrectly() {
        // Given
        val record = SkinTemperatureRecord(
            startTime = Instant.ofEpochMilli(TEST_START_TIME),
            endTime = Instant.ofEpochMilli(TEST_END_TIME),
            startZoneOffset = TEST_ZONE_OFFSET,
            endZoneOffset = TEST_ZONE_OFFSET,
            metadata = Metadata.manualEntry(),
            deltas = listOf(
                SkinTemperatureRecord.Delta(
                    time = Instant.ofEpochMilli(TEST_DELTA_TIME_1),
                    delta = TemperatureDelta.celsius(TEST_TEMPERATURE_DELTA_CELSIUS_1),
                ),
            ),
            baseline = null,
            measurementLocation = SkinTemperatureRecord.MEASUREMENT_LOCATION_TOE,
        )

        // When
        val result = record.toDto()

        // Then
        result.baselineCelsius shouldBe null
        result.measurementLocation shouldBe SkinTemperatureMeasurementLocationDto.TOE
    }

    @Test
    @DisplayName(
        "GIVEN SkinTemperatureDeltaSeriesRecordDto with null baseline → WHEN toHealthConnect → THEN creates correctly",
    )
    fun whenDtoToHealthConnectWithNullBaseline_thenCreatesCorrectly() {
        // Given
        val dto = SkinTemperatureDeltaSeriesRecordDto(
            id = TEST_ID,
            startTime = TEST_START_TIME,
            endTime = TEST_END_TIME,
            startZoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            endZoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            metadata = MetadataDto(
                dataOrigin = "com.example.app",
                recordingMethod = RecordingMethodDto.MANUAL_ENTRY,
                deviceType = DeviceTypeDto.PHONE,
            ),
            samples = listOf(
                SkinTemperatureDeltaSampleDto(
                    time = TEST_DELTA_TIME_1,
                    temperatureDeltaCelsius = TEST_TEMPERATURE_DELTA_CELSIUS_1,
                ),
            ),
            baselineCelsius = null,
            measurementLocation = SkinTemperatureMeasurementLocationDto.UNKNOWN,
        )

        // When
        val result = dto.toHealthConnect()

        // Then
        result.baseline shouldBe null
        result.measurementLocation shouldBe SkinTemperatureRecord.MEASUREMENT_LOCATION_UNKNOWN
    }

    @Test
    @DisplayName(
        "GIVEN SkinTemperatureDeltaSeriesRecordDto with null measurementLocation → WHEN toHealthConnect → THEN defaults to UNKNOWN",
    )
    fun whenDtoToHealthConnectWithNullMeasurementLocation_thenDefaultsToUnknown() {
        // Given
        val dto = SkinTemperatureDeltaSeriesRecordDto(
            id = TEST_ID,
            startTime = TEST_START_TIME,
            endTime = TEST_END_TIME,
            startZoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            endZoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            metadata = MetadataDto(
                dataOrigin = "com.example.app",
                recordingMethod = RecordingMethodDto.MANUAL_ENTRY,
                deviceType = DeviceTypeDto.PHONE,
            ),
            samples = emptyList(),
            baselineCelsius = null,
            measurementLocation = null,
        )

        // When
        val result = dto.toHealthConnect()

        // Then
        result.measurementLocation shouldBe SkinTemperatureRecord.MEASUREMENT_LOCATION_UNKNOWN
    }
}

File: packages/health_connector_hc_android/test/unit_tests/src/mappers/health_record_mappers/temperature/skin_temperature_delta_series_record_mapper_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/temperature/skin_temperature_delta_series_record_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart';

import '../../../../utils/fake_data.dart';

void main() {
  group(
    'SkinTemperatureDeltaSeriesRecordMapper',
    () {
      group(
        'SkinTemperatureDeltaSeriesRecordToDto',
        () {
          test(
            'converts SkinTemperatureDeltaSeriesRecord to '
            'SkinTemperatureDeltaSeriesRecordDto',
            () {
              final samples = [
                SkinTemperatureDeltaSample(
                  time: FakeData.fakeStartTime,
                  temperatureDelta: const Temperature.celsius(0.2),
                ),
                SkinTemperatureDeltaSample(
                  time: FakeData.fakeStartTime.add(const Duration(minutes: 5)),
                  temperatureDelta: const Temperature.celsius(-0.1),
                ),
              ];

              final record = SkinTemperatureDeltaSeriesRecord.internal(
                id: HealthRecordId(FakeData.fakeId),
                startTime: FakeData.fakeStartTime,
                endTime: FakeData.fakeEndTime,
                startZoneOffsetSeconds:
                    FakeData.fakeStartTime.timeZoneOffset.inSeconds,
                endZoneOffsetSeconds:
                    FakeData.fakeEndTime.timeZoneOffset.inSeconds,
                metadata: Metadata.internal(
                  dataOrigin: const DataOrigin(FakeData.fakeDataOrigin),
                  recordingMethod: RecordingMethod.manualEntry,
                  clientRecordVersion: 1,
                  device: const Device(type: DeviceType.phone),
                ),
                samples: samples,
                baseline: const Temperature.celsius(36.5),
                measurementLocation: SkinTemperatureMeasurementLocation.wrist,
              );

              final dto = record.toDto();

              expect(dto.id, FakeData.fakeId);
              expect(
                dto.startTime,
                FakeData.fakeStartTime.millisecondsSinceEpoch,
              );
              expect(dto.endTime, FakeData.fakeEndTime.millisecondsSinceEpoch);
              expect(
                dto.startZoneOffsetSeconds,
                FakeData.fakeStartTime.timeZoneOffset.inSeconds,
              );
              expect(
                dto.endZoneOffsetSeconds,
                FakeData.fakeEndTime.timeZoneOffset.inSeconds,
              );
              expect(dto.baselineCelsius, 36.5);
              expect(dto.samples.length, 2);
              expect(dto.samples[0].temperatureDeltaCelsius, 0.2);
              expect(dto.samples[1].temperatureDeltaCelsius, -0.1);
            },
          );

          test(
            'converts SkinTemperatureDeltaSeriesRecord with null '
            'baseline to DTO',
            () {
              final samples = [
                SkinTemperatureDeltaSample(
                  time: FakeData.fakeStartTime,
                  temperatureDelta: const Temperature.celsius(0.3),
                ),
              ];

              final record = SkinTemperatureDeltaSeriesRecord.internal(
                id: HealthRecordId(FakeData.fakeId),
                startTime: FakeData.fakeStartTime,
                endTime: FakeData.fakeEndTime,
                startZoneOffsetSeconds:
                    FakeData.fakeStartTime.timeZoneOffset.inSeconds,
                endZoneOffsetSeconds:
                    FakeData.fakeEndTime.timeZoneOffset.inSeconds,
                metadata: Metadata.internal(
                  dataOrigin: const DataOrigin(FakeData.fakeDataOrigin),
                  recordingMethod: RecordingMethod.manualEntry,
                  clientRecordVersion: 1,
                  device: const Device(type: DeviceType.phone),
                ),
                samples: samples,
              );

              final dto = record.toDto();

              expect(dto.baselineCelsius, isNull);
            },
          );
        },
      );

      group(
        'SkinTemperatureDeltaSeriesRecordDtoToDomain',
        () {
          test(
            'converts SkinTemperatureDeltaSeriesRecordDto to '
            'SkinTemperatureDeltaSeriesRecord',
            () {
              final dto = SkinTemperatureDeltaSeriesRecordDto(
                id: FakeData.fakeId,
                startTime: FakeData.fakeLocalStartTime.millisecondsSinceEpoch,
                endTime: FakeData.fakeLocalEndTime.millisecondsSinceEpoch,
                startZoneOffsetSeconds:
                    FakeData.fakeStartTime.timeZoneOffset.inSeconds,
                endZoneOffsetSeconds:
                    FakeData.fakeEndTime.timeZoneOffset.inSeconds,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.activelyRecorded,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.phone,
                ),
                samples: [
                  SkinTemperatureDeltaSampleDto(
                    time: FakeData.fakeStartTime.millisecondsSinceEpoch,
                    temperatureDeltaCelsius: 0.2,
                  ),
                  SkinTemperatureDeltaSampleDto(
                    time: FakeData.fakeStartTime
                        .add(const Duration(minutes: 5))
                        .millisecondsSinceEpoch,
                    temperatureDeltaCelsius: -0.1,
                  ),
                ],
                baselineCelsius: 36.5,
                measurementLocation:
                    SkinTemperatureMeasurementLocationDto.wrist,
              );

              final record = dto.toDomain();

              expect(record.id.value, FakeData.fakeId);
              expect(record.startTime, FakeData.fakeStartTime);
              expect(record.endTime, FakeData.fakeEndTime);
              expect(record.baseline?.inCelsius, 36.5);
              expect(record.samples.length, 2);
              expect(record.samples[0].temperatureDelta.inCelsius, 0.2);
              expect(record.samples[1].temperatureDelta.inCelsius, -0.1);
              expect(
                record.measurementLocation,
                SkinTemperatureMeasurementLocation.wrist,
              );
            },
          );

          test(
            'converts SkinTemperatureDeltaSeriesRecordDto with null '
            'baseline to domain',
            () {
              final dto = SkinTemperatureDeltaSeriesRecordDto(
                id: FakeData.fakeId,
                startTime: FakeData.fakeLocalStartTime.millisecondsSinceEpoch,
                endTime: FakeData.fakeLocalEndTime.millisecondsSinceEpoch,
                startZoneOffsetSeconds: 0,
                endZoneOffsetSeconds: 0,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.manualEntry,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.phone,
                ),
                samples: [
                  SkinTemperatureDeltaSampleDto(
                    time: FakeData.fakeStartTime.millisecondsSinceEpoch,
                    temperatureDeltaCelsius: 0.3,
                  ),
                ],
              );

              final record = dto.toDomain();

              expect(record.baseline, isNull);
              expect(
                record.measurementLocation,
                SkinTemperatureMeasurementLocation.unknown,
              );
            },
          );

          test(
            'converts SkinTemperatureDeltaSeriesRecordDto with null '
            'id to domain with none id',
            () {
              final dto = SkinTemperatureDeltaSeriesRecordDto(
                startTime: FakeData.fakeLocalStartTime.millisecondsSinceEpoch,
                endTime: FakeData.fakeLocalEndTime.millisecondsSinceEpoch,
                startZoneOffsetSeconds: 0,
                endZoneOffsetSeconds: 0,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.manualEntry,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.phone,
                ),
                samples: [],
                baselineCelsius: 37.0,
                measurementLocation:
                    SkinTemperatureMeasurementLocationDto.finger,
              );

              final record = dto.toDomain();

              expect(record.id, HealthRecordId.none);
            },
          );
        },
      );
    },
  );
}
